llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp

```diff
@@ -210,36 +219,105 @@ void InterBlockScheduling::leaveFunction() {
-                      << CurrentBlock->FixPoint.NumIters << "\n");
-}
+                      << CurrentBlockState->kindAsString() << " FixPointIter="
+                      << CurrentBlockState->FixPoint.NumIters
+                      << " II=" << CurrentBlockState->FixPoint.II << "\n");
+}
+namespace {
+/// This implements the interface to the postpipeliner to extract the
+/// prologue, steady state and epilogue of the pipelined loop
+/// Each of these code segments is collected in TimedRegion, and then copied
+/// into the appropriate blockstate region.
+/// TimedRegion is built one bundle at the time
+class PipelineExtractor : public PipelineScheduleVisitor {
+  BlockState &Loop;
+  BlockState *Prologue = nullptr;
+  BlockState *Epilogue = nullptr;
+  MachineBundle CurrentBundle;
+  std::vector<MachineBundle> TimedRegion;
+  bool InLoop = false;
+
+  void startPrologue() override {
+    // Nothing at this time, but let's keep the override around
+  }
+  void startLoop() override {
+    Prologue->BottomInsert = TimedRegion;
+    TimedRegion.clear();
+    InLoop = true;
+  }
+  void startEpilogue() override {
+    Loop.getTop().Bundles = TimedRegion;
+    TimedRegion.clear();
+    InLoop = false;
+  }
+  void finish() override {
+    Epilogue->TopInsert = TimedRegion;
+    TimedRegion.clear();
+  }
+  void startBundle() override { CurrentBundle.clear(); }
+  void addToBundle(MachineInstr *MI) override {
+    // We re-emit the original instructions into the loop body.
+    // Prologue and epilogue obtain copies.
+    MachineInstr *ToBeEmitted =
+        InLoop ? MI : Loop.TheBlock->getParent()->CloneMachineInstr(MI);
+    CurrentBundle.add(ToBeEmitted);
+  }
+  void endBundle() override { TimedRegion.emplace_back(CurrentBundle); }
+
+public:
+  PipelineExtractor(InterBlockScheduling &InterBlock, BlockState &BS,
+                    const AIEBaseInstrInfo &TII)
+      : Loop(BS), CurrentBundle(TII.getFormatInterface()) {
+    MachineBasicBlock *LoopBlock = Loop.TheBlock;
+    for (auto *P : LoopBlock->predecessors()) {
+      if (P == LoopBlock) {
+        continue;
+      }
+      Prologue = &InterBlock.getBlockState(P);
+    }
+    for (auto *S : LoopBlock->successors()) {
+      if (S == LoopBlock) {
+        continue;
+      }
+      Epilogue = &InterBlock.getBlockState(S);
+    }
+  }
+};
 
+} // namespace
 bool InterBlockScheduling::leaveBlock() {
   DEBUG_BLOCKS(dbgs() << "  << leaveBlock "
-                      << CurrentBlock->TheBlock->getNumber() << "\n");
+                      << CurrentBlockState->TheBlock->getNumber() << "\n");
   // After scheduling a basic block, check convergence to determine which block
   // to schedule next and with what parameters
-  auto &BS = *CurrentBlock;
-  if (BS.Kind == BlockType::Loop && !updateFixPoint(BS)) {
-    BS.FixPoint.NumIters++;
+  auto &BS = *CurrentBlockState;
+  switch (updateFixPoint(BS)) {
+  case SchedulingStage::SchedulingNotConverged:
```

Comment:
I guess this is just for completeness, but we will have crashed before checking `SchedulingStage::SchedulingNotConverged` here?

Reply:
Correct. Hence that I wrote 'may not be observable.` at the enum definition.